src/farbfeld.rs

```diff
@@ -63,44 +67,98 @@ impl<R: Read> FarbfeldReader<R> {
             width: read_dimm(&mut inner)?,
             height: read_dimm(&mut inner)?,
             inner,
+            current_offset: 0,
             cached_byte: None,
         })
     }
 }
 
 impl<R: Read> Read for FarbfeldReader<R> {
     fn read(&mut self, mut buf: &mut [u8]) -> io::Result<usize> {
-        fn consume_channel<R: Read>(from: &mut R, to: &mut [u8]) -> io::Result<()> {
-            let mut ibuf = [0u8; 2];
-            from.read_exact(&mut ibuf)?;
-            NativeEndian::write_u16(to, BigEndian::read_u16(&ibuf));
-            Ok(())
-        }
-
         let mut bytes_written = 0;
         if let Some(byte) = self.cached_byte.take() {
             buf[0] = byte;
             buf = &mut buf[1..];
             bytes_written = 1;
+            self.current_offset += 1;
         }
 
         if buf.len() == 1 {
-            let mut obuf = [0u8; 2];
-            consume_channel(&mut self.inner, &mut obuf)?;
-            buf[0] = obuf[0];
-            self.cached_byte = Some(obuf[1]);
+            buf[0] = cache_byte(&mut self.inner, &mut self.cached_byte)?;
             bytes_written += 1;
+            self.current_offset += 1;
         } else {
             for channel_out in buf.chunks_exact_mut(2) {
                 consume_channel(&mut self.inner, channel_out)?;
                 bytes_written += 2;
+                self.current_offset += 2;
             }
         }
 
         Ok(bytes_written)
     }
 }
 
+impl<R: Read + Seek> Seek for FarbfeldReader<R> {
+    fn seek(&mut self, pos: SeekFrom) -> io::Result<u64> {
+        fn parse_offset(original_offset: u64, end_offset: u64, pos: SeekFrom) -> Option<i64> {
+            match pos {
+                SeekFrom::Start(off) =>
+                    i64::try_from(off).ok()?.checked_sub(i64::try_from(original_offset).ok()?),
+                SeekFrom::End(off) =>
+                    if off < i64::try_from(end_offset).unwrap_or(i64::MAX) {
+                        None
+                    } else {
+                        Some(i64::try_from(end_offset.checked_sub(original_offset)?).ok()? + off)
+                    },
+                SeekFrom::Current(off) =>
+                    if off < i64::try_from(original_offset).unwrap_or(i64::MAX) {
+                        None
+                    } else {
+                        Some(off)
+                    },
+            }
+        }
+
+        let original_offset = self.current_offset;
+        let end_offset = self.width as u64 * self.height as u64 * 2;
+        let offset_from_current = parse_offset(original_offset, end_offset, pos)
+            .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidInput, "invalid seek to a negative or overflowing position"))?;
+
+        // TODO: convert to seek_relative() once that gets stabilised
```

Comment:
When this is merged, could you open an issue for this, with the label `blocked-msrv`?

Reply:
Will do!